File: Irrgarten/src/irrgarten/Game.java
package irrgarten;
import java.util.ArrayList;

/**
 *
 * @author Sergio Salvador Gil
 */
public class Game {
    
    private static final int MAX_ROUNDS=10;
    private int currentPlayerIndex;
    
    
    private String log;
    
    private Labyrinth labyrinth;
    private ArrayList<Monster> monsters;
    private Player currentPlayer;
    private ArrayList<Player> players;
    
    // Métodos públicos de la clase Game
    
    public Game(int nplayers){
        
        // Generar los jugadores
        
        this.players = new ArrayList<>();
        Dice Dice = new Dice();
        for (int i = 0; i < nplayers; i++){
            Player actual = new Player((char)(i+'0'),Dice.randomIntelligence(), Dice.randomStrength());
            players.add(actual);
        }
        
        // Decidir quién empieza
        currentPlayerIndex=Dice.whoStarts(nplayers);
        currentPlayer=players.get(currentPlayerIndex);
        
        monsters = new ArrayList();
        labyrinth = new Labyrinth(7,5,2,3);
        
        log="";
        
        configureLabyrinth();
        labyrinth.spreadPlayers(players);
        
    }
    
    public boolean finished(){
        return labyrinth.haveAWinner();
    }
    
    public boolean nextStep(Directions preferredDirection){
        log ="";
        boolean dead = currentPlayer.dead();
        
        if (!dead){
            
            Directions direction = this.actualDirection(preferredDirection);
            if (direction != preferredDirection){
                logPlayerNoOrders();
            }
            Monster monster = labyrinth.putPlayer(direction, currentPlayer);
            
            if (monster == null){
                this.logNoMonster();
            }
            else{
                GameCharacter winner = combat(monster);
                this.manageReward(winner);
            }
            
        }
        else this.manageResurrection();
        
        boolean endGame = this.finished();
        
        if (!endGame){
            this.nextPlayer();
        }
        
        return endGame;
    }
    
    public GameState getGameState(){
        String playersStr="";
        String monstersStr="";
        
        for (int i = 0; i < players.size(); i++){
            playersStr+=players.get(i).toString()+"\n";
        }
        for (int i = 0; i < monsters.size(); i++){
            monstersStr+=monsters.get(i).toString()+"\n";
        }
        
        GameState estado = new GameState(labyrinth.toString(), playersStr, 
        monstersStr, currentPlayerIndex, this.finished(), log);
        return estado;
    }
    //Métodos privados de la clase Game
    
    private void configureLabyrinth(){
        Monster monstruo1= new Monster("Netanyahu", Dice.randomIntelligence(), Dice.randomStrength());
        Monster monstruo2= new Monster("Bin Laden", Dice.randomIntelligence(), Dice.randomStrength());
        Monster monstruo3= new Monster("Khamenei", Dice.randomIntelligence(), Dice.randomStrength());
        
        labyrinth.addMonster(3,3,monstruo1);
        labyrinth.addMonster(1,4,monstruo2);
        labyrinth.addMonster(5,2,monstruo3);
        monsters.add(monstruo1);
        monsters.add(monstruo2);
        monsters.add(monstruo3);
        
        labyrinth.addBlock(Orientation.HORIZONTAL, 2,0,3);
        labyrinth.addBlock(Orientation.HORIZONTAL, 1,3,1);
    }
    
    private void nextPlayer(){
        if (currentPlayerIndex==players.size()-1){
            currentPlayerIndex=0;
        }
       
        else  currentPlayerIndex++;
        
        currentPlayer = players.get(currentPlayerIndex);
    }
    
    private Directions actualDirection(Directions preferredDirection){
        int row = currentPlayer.getRow();
        int col = currentPlayer.getCol();
        
        ArrayList<Directions> validMoves=labyrinth.validMoves(row,col);
        
        return currentPlayer.move(preferredDirection, validMoves);
    }
    
    private GameCharacter combat(Monster monster){
        int rounds = 0;
        GameCharacter winner = GameCharacter.PLAYER;
        
        float playerAttack=currentPlayer.attack();
        boolean lose = monster.defend(playerAttack);
        
        while (!lose && (rounds < MAX_ROUNDS)){
            winner = GameCharacter.MONSTER;
            rounds++;
            
            float monsterAttack = monster.attack();
            lose = currentPlayer.defend(monsterAttack);
            
            if (!lose){
                playerAttack = currentPlayer.attack();
                winner = GameCharacter.PLAYER;
                lose = monster.defend(playerAttack);
            }
        }
        this.logRounds(rounds, MAX_ROUNDS);
        return winner;
    }
    
    private void manageReward(GameCharacter winner){
        
        if (winner==GameCharacter.PLAYER){
            currentPlayer.receiveReward();
            this.logPlayerWon();
        }
        else this.logMonsterWon();
    }
    
    private void manageResurrection(){
        Dice Dice = new Dice();
        
        boolean resurrect = Dice.resurrectPlayer();
        
        if (resurrect){
            currentPlayer.resurrect();
            this.logResurrected();
        }
        else this.logPlayerSkipTurn();
    }
    
    private void logPlayerWon(){
        log+="El jugador ha ganado el combate\n";
    }
    
    private void logMonsterWon(){
        log+="El monstruo ha ganado el combate\n";
    }
    
    private void logResurrected(){
        log+="El jugador ha resucitado\n";
    }
    
    private void logPlayerSkipTurn(){
        log+="El jugador ha perdido su turno por estar muerto\n";
    }
    
    private void logPlayerNoOrders(){
        log+="El jugador no ha podido cumplir las órdenes\n";
    }
    
    private void logNoMonster(){
        log+="El jugador se ha movido a una celda vacía o no le ha sido posible moverse\n";
    }
    
    private void logRounds(int rounds, int max){
        log+="Se han producido "+rounds+" de "+max+" rondas de combate\n";
    }
    
}

File: rb/Irrgarten/Game.rb
#encoding:UTF-8

module Irrgarten
    class Game

        MAX_ROUNDS=10

        def initialize(n_players)
            dado = Dice.new

            @log=""
            @labyrinth = Labyrinth.new(4,4,2,3)
            @players = Array.new
            @monsters = Array.new

            n_players.times do |i|
                actual = Player.new(i, dado.random_intelligence, dado.random_strength)
                @players << actual
            end
            @current_player_index=dado.who_starts(n_players)
            @current_player=@players[@current_player_index]

            configure_labyrinth
            #@labyrinth.spread_players(@players)
        end

        def finished
            return @labyrinth.have_a_winner
        end

        def next_step(preferred_direction)
            log = ""
            boolean dead = @current_player.dead
            if (!dead)
                direction = actual_direction(preferred_direction)
                if (direction != preferred_direction)
                    log_player_no_orders
                end
                monster = @labyrinth.put_player(direction, @current_player)
                if (monster == nil)
                    log_no_monster
                else
                    winner = combat(monster)
                    manage_reward(winner)
                end
            else
                manage_resurrection
            end
            end_game=finished
            if (!end_game)
                next_player
            end
            end_game
        end

        def get_game_state

            laberinto = @labyrinth.to_string
            jugadores = @players.map { |p| p.to_s }.join("\n")
            monstruos = @monsters.map { |m| m.to_s }.join("\n")
            estado = GameState.new(laberinto, jugadores, monstruos, @current_player, finished, @log)

            return estado
        end


        private


        def configure_labyrinth
            dado = Dice.new
            monstruo1 = Monster.new("Netanyahu", dado.random_intelligence, dado.random_strength)
            monstruo2 = Monster.new("Bin Laden", dado.random_intelligence, dado.random_strength)
            monstruo3 = Monster.new("Khamenei", dado.random_intelligence, dado.random_strength)

            @labyrinth.add_monster(3,3,monstruo1)
            @labyrinth.add_monster(3,1,monstruo2)
            @labyrinth.add_monster(3,0,monstruo3)
            
        end

        def next_player 
            if (@current_player_index==@players.size-1)
                @current_player_index=0
            else
                @current_player_index+=1
            end

            @current_player = @players[@current_player_index]
        end

        def actual_direction(preferredDirection)
            current_row=@current_player.get_row()
            current_col=@current_player.get_col()

            valid_moves = labyrinth.valid_moves(current_row,current_col)
            output = move(preferredDirection,valid_moves)
            output
        end

        def combat(monster)
            rounds = 0
            winner = GameCharacter::PLAYER

            player_attack=@current_player.attack
            lose = monster.defend(player_attack)

            while (!lose && rounds < MAX_ROUNDS) do
                rounds+=1
                winner=GameCharacter::MONSTER
                monster_attack=monster.attack
                lose = @current_player.defend(monster_attack)

                if (!lose)
                    winner=GameCharacter::PLAYER
                    player_attack=@current_player.attack
                    lose = monster.defend(player_attack)
                end
            end
            log_rounds
            winner
        end

        def manage_reward(winner)
            if (winner==GameCharacter::PLAYER)
                @current_player.receive_reward
                log_player_won
            else
                log_monster_won
            end
        end

        def manage_resurrection
            dado = Dice.new
            resurrect = dado.resurrect_player
            if (resurrect)
                @current_player.resurrect
                log_resurrected
            else
                log_player_skip_turn
            end
        end

        def log_player_won
            @log+="El jugador ha ganado el combate\n";
        end

        def log_monster_won
            @log+="El monstruo ha ganado el combate\n";
        end

        def log_resurrected
            @log+="El jugador ha resucitado\n";
        end

        def log_player_skip_turn
            @log+="El jugador ha perdido su turno por estar muerto\n";
        end

        def log_player_no_orders
            @log+="El jugador no ha podido cumplir las órdenes\n";
        end

        def log_no_monster
            @log+="El jugador se ha movido a una celda vacía o no le ha sido posible moverse\n";
        end

        def log_rounds(rounds,max)
            log+="Se han producido #{rounds} de #{max} rondas de combate\n"
        end

    end
end